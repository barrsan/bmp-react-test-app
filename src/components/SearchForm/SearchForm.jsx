import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { throttle } from 'throttle-debounce';


class SearchForm extends PureComponent {
  static propTypes = {
    setSearchString: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props);
    this.state = {
      value: '',
    };
    this.searchThrottled = throttle(500, this.search);
  }

  search = (searchString) => {
    const { setSearchString } = this.props;
    setSearchString({
      searchString,
    });
  }

  handleChange = (e) => {
    e.persist();
    this.setState(prevState => (
      { ...prevState, value: e.target.value }
    ), () => {
      const { value } = this.state;
      this.searchThrottled(value);
    });
  }

  render() {
    const { value } = this.state;
    return (
      <div className="search-form form-group">
        <input
          className="form-control"
          placeholder="Поиск по названию"
          value={value}
          onChange={this.handleChange}
        />
      </div>
    );
  }
}


export default SearchForm;
