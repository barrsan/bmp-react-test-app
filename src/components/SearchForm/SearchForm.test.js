import React from 'react';
import { shallow } from 'enzyme';
import SearchForm from './SearchForm';


describe('SearchForm component', () => {
  const mockSetSearchString = jest.fn();
  const initialState = {
    value: '',
  };
  const props = {
    setSearchString: mockSetSearchString,
  };

  const wrapper = shallow(<SearchForm {...props} />);

  test('renders correctly', () => expect(wrapper).toMatchSnapshot());

  test('renders without crashing', () => {
    expect(wrapper).toHaveLength(1);
  });

  test('initialize with initial state', () => {
    expect(wrapper.state()).toEqual(initialState);
  });


  describe('When typing into search input', () => {
    const value = 'test';

    beforeEach(() => {
      wrapper.find('input').simulate('change', {
        persist: () => {},
        target: {
          value,
        },
      });
    });

    test('updates value field in state', () => {
      expect(wrapper.state().value).toEqual(value);
    });

    test('calls the props.setSearchString', () => {
      setTimeout(() => {
        expect(mockSetSearchString).toBeCalled();
      }, 500);
    });
  });

  afterAll(() => {
    wrapper.setState(initialState);
  });
});
