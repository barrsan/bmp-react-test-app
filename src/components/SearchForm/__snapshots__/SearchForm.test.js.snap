// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchForm
    setSearchString={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <input
        className="form-control"
        onChange={[Function]}
        placeholder="Поиск по названию"
        value=""
/>,
      "className": "search-form form-group",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "form-control",
        "onChange": [Function],
        "placeholder": "Поиск по названию",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <input
          className="form-control"
          onChange={[Function]}
          placeholder="Поиск по названию"
          value=""
/>,
        "className": "search-form form-group",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-control",
          "onChange": [Function],
          "placeholder": "Поиск по названию",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
